rust/kernel/clk.rs

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Clock subsystem
+//!
+//! C header: [`include/linux/clk.h`](../../../../include/linux/clk.h)
+
+use crate::{
+    bindings,
+    device::Device,
+    error::{from_kernel_err_ptr, Error, Result},
+    str::CStr,
+};
+
+/// Represents `struct clk *`
+pub struct Clk(*mut bindings::clk);
+
+impl Clk {
+    /// returns value of rate field of `struct clk`
+    pub fn get_rate(&self) -> u32 {
+        // SAFETY: self object contains aquired, valid pointer to `struct clk`
+        unsafe { bindings::clk_get_rate((*self).0) }
+    }
+}
+
+/// clk_prepare_enable helps cases using clk_enable in non-atomic context
+pub fn prepare_enable(clk: &mut Clk) -> Result {
+    // SAFETY: clk is an invariant containing non-null pointer to `struct clk`
+    let ret = unsafe { bindings::clk_prepare_enable((*clk).0) };
+    if ret < 0 {
+        return Err(Error::from_kernel_errno(ret));
+    }
+
+    Ok(())
+}
+
+/// Lookup and obtain a managed reference to a clock producer
+/// - `dev` - device for clock "consumer"
+/// - `id` - clock consumer ID
+pub fn devm_clk_get(dev: &mut impl Device, id: &'static CStr) -> Result<Clk> {
+    // SAFETY: FFI call, dev must contain valid, non-null `struct dev` ptr
```

Comment:
In general, "must" shouldn't appear in `SAFETY` comments, because they are explanations of why something is sound, not contracts.

Here, for instance, what guarantees `dev.mut_raw()` returns a valid pointer? An `impl` could return a NULL pointer, for intsance.

Reply:
So in this case a `Device` structure needs to created that must fulfill a safety contract so that it's invariant will be valid?

Reply:
Yeah, but note that if you need that somebody fulfills a safety contract when implementing a trait, you need to mark the trait as `unsafe` (or perhaps mark some of its functions as unsafe).

Put another way, one could implement the trait as:

```rust
impl Device for MyDevice {
    fn mut_raw(&mut self) -> *mut BindingsDevice {
        core::ptr::null_mut()
    }
    // ...
}
```

and then call:

```rust
let my_device: MyDevice = ...;
devm_clk_get(my_device.mut_raw(), "...");
```

and they would be passing a NULL pointer to the C side without unsafe code.

Now, it may be the case that the C side in this case handles the NULL case gracefully, but `mut_raw` could also return any other non-NULL invalid value, too.

Reply:
Yeah, that explains a lot. I had not deduced it from the code before. Thank you!
This could be part of some documentation I belive.

The merged commit https://github.com/Rust-for-Linux/linux/commit/6f94f6a76296cc8af6ce14bb7d43dc67a95b0ef6 provides the solution already so I will rebase the code based on it.

Reply:
You're welcome!